Splice: Complete tx_abort implementation

ddustin opened the pull request:
Adds support for `tx_abort` including sending, receiving, and ACK'ing.

When `tx_abort` is confirmed in `channeld` we notify `lightningd` and shutdown. `lightningd` then restarts the channeld daemon.

This is all done without breaking the connection and special care is made to connect the daemon to the existing connection through a `connectd` fd.

ksedgwic commented:
@ddustin should we try this out (integration tests) w/ our splicing branch?  Our branch is not quite released yet, but is close ...

ddustin commented:
> @ddustin should we try this out (integration tests) w/ our splicing branch?  Our branch is not quite released yet, but is close ...

It's not quite ready for testing yet

ksedgwic commented:
@devrandom review beg

ksedgwic commented:
@ddustin I think I'm making some assumptions, want to make sure they are correct:
- `tx_abort` is used to cancel a in-progress interactive splice tx negotiation
- the semantics are the same as dual-funding

I found the dual-funding docs and 5 comments on https://github.com/lightning/bolts/pull/863, is there anything additional I should be looking at?

ddustin commented:
> @ddustin I think I'm making some assumptions, want to make sure they are correct:
> 
> - `tx_abort` is used to cancel a in-progress interactive splice tx negotiation
> 
> - the semantics are the same as dual-funding
> 
> 
> 
> I found the dual-funding docs and 5 comments on https://github.com/lightning/bolts/pull/863, is there anything additional I should be looking at?

The spec for `tx_abort` is in dual and it's essentially the same -- though the implementation for splice abort is in channeld.c while dual's abort is in dualopend.c

cdecker commented:
Hi @ddustin, just going through all the open PRs that may make it into v24.02, and this one looks good for inclusion as far as I can see. Can you share your view on the current status, and whether we can include it or not? If included, we may do 1-2 rounds of final reviews and merge it, otherwise I'll untag it for the milestone, until it is ready for inclusion ^^

ddustin commented:
> Hi @ddustin, just going through all the open PRs that may make it into v24.02, and this one looks good for inclusion as far as I can see. Can you share your view on the current status, and whether we can include it or not? If included, we may do 1-2 rounds of final reviews and merge it, otherwise I'll untag it for the milestone, until it is ready for inclusion ^^

Hi @cdecker, it would be great to get this included in the next release! I believe it is definitely ready for inclusion.

This should all only apply under the experimental-splicing flag making it a safer change.

In some manual tests aggressively restarting channeld (which tx_abort does) reveals some passed file descriptor issues. I believe this is revealing an existing issue and not related to this PR.

If correct this should enable us to add flaky fd passing to the CI. To build the test however I need PR #6980 merged too as the test is dependent on both.

cdecker commented:
Ok, got it. I quite like #6980, but the discussion doesn't look finished yet. Am I good to take on this PR and shepherd it through CI, even if #6980 may not make it through in this release? The next release would be 24.05, so not a huge delay, but I prefer pulling in as much as possible to keep a clean worktable.

ddustin commented:
> Ok, got it. I quite like #6980, but the discussion doesn't look finished yet. Am I good to take on this PR and shepherd it through CI, even if #6980 may not make it through in this release? The next release would be 24.05, so not a huge delay, but I prefer pulling in as much as possible to keep a clean worktable.

Yeah that works 👍, thanks!

rustyrussell commented on lightningd/channel_control.c:
```diff
@@ -284,6 +284,95 @@ static void handle_splice_feerate_error(struct lightningd *ld,
+								     &channel->cid)));
+		subd_send_fd(ld->connectd, fds[1]);
+		log_info(channel->log, "Sent the peer fd to channeld");
+	}
+	else {
```
Weird linebreak

rustyrussell commented on channeld/channeld.c:
```diff
@@ -1625,6 +1631,197 @@ static void send_revocation(struct peer *peer,
 	peer_write(peer->pps, take(msg));
 }
 
+static struct inflight *last_inflight(struct peer *peer)
```
const struct peer *

rustyrussell commented on channeld/channeld.c:
```diff
@@ -1625,6 +1631,197 @@ static void send_revocation(struct peer *peer,
+	return NULL;
+}
+
+static size_t last_inflight_index(struct peer *peer)
```
here too...

rustyrussell commented on channeld/channeld.c:
```diff
@@ -1625,6 +1631,197 @@ static void send_revocation(struct peer *peer,
+	return tal_count(peer->splice_state->inflights) - 1;
+}
+
+static u32 find_channel_funding_input(struct wally_psbt *psbt,
+				      struct bitcoin_outpoint *funding)
```
Both of these can be const...

rustyrussell commented on channeld/channeld.c:
```diff
@@ -1625,6 +1631,197 @@ static void send_revocation(struct peer *peer,
+	splice_funding_index = find_channel_funding_input(inflight->psbt,
+							  &peer->channel->funding);
+	sigs_needed = 0;
+	for (i = 0; i < inflight->psbt->num_inputs; i++) {
```
for (size_t i.... since you don't use the i var outside the loop?

rustyrussell commented on channeld/channeld.c:
```diff
@@ -1625,6 +1631,197 @@ static void send_revocation(struct peer *peer,
+	struct channel_id channel_id;
+	u8 *reason;
+
+	if (!msg || fromwire_peektype(msg) != WIRE_TX_ABORT)
```
FWIW, fromwire_peektype returns -1 on NULL, so it's safe.

rustyrussell reviewed (changes requested):
Trivial changes mainly.


But last commit should be the first: that means tests always pass!

cdecker commented:
I tried rebasing this PR, but managed to somehow break it. It now has an undefined symbol `towire_connectd_peer_final_msg` and I can't find a trace of that message anywhere.

ddustin commented:
> I tried rebasing this PR, but managed to somehow break it. It now has an undefined symbol `towire_connectd_peer_final_msg` and I can't find a trace of that message anywhere.

Ah looks like commit db6f0da3b349d89892e8d96012c3ab184735593b removed that routine. Should be a simple update to use the new method.

ddustin commented:
> Trivial changes mainly.
> 
> But last commit should be the first: that means tests always pass!

Thanks for the review!

I addressed these changes, rebased master, and updated it to work with the latest master.

@cdecker once it passes CI it should be good to go
